File: go/p36.go
package main

import "fmt"

func getDigits(n, b int) []int {
    digits := []int{}
    for n > 0 {
        digits = append(digits, n % b)
        n /= b
    }
    return digits
}

func isPalindrome(digits []int) bool {
    left := 0
    right := len(digits) - 1

    for left < right {
        if digits[left] != digits[right] {
            return false
        }

        left++
        right--
    }

    return true
}

func main() {
    sum := 0
    for i := 1; i != 1000000; i++ {
        if isPalindrome(getDigits(i, 10)) && isPalindrome(getDigits(i, 2)) {
            sum += i
        }
    }
    fmt.Println(sum)
}

File: cpp/p36.cpp
#include <iostream>
using namespace std;

bool PalindromeInBase(int x, int base)
{
    int digits[32];
    bool couldBePalindrome = true;
    
    int pos = 0;
    while (x != 0)
    {
        digits[pos++] = x % base;
        x /= base;
    }
    
    int left = 0;
    int right = pos - 1;
    while (left < right && couldBePalindrome)
        couldBePalindrome = digits[left++] == digits[right--];
    
    return couldBePalindrome;
}

int main()
{
    int sum = 0;
    for (int i = 1; i != 1000000; ++i)
        if (PalindromeInBase(i, 10) && PalindromeInBase(i, 2))
            sum += i;
    
    cout << sum << endl;
    
    return 0;
}
